sdk/eventhubs/azure_messaging_eventhubs/src/common/connection_manager.rs

```diff
@@ -0,0 +1,150 @@
+// Copyright (c) Microsoft Corporation. All Rights reserved
+// Licensed under the MIT license.
+
+use crate::common::user_agent::get_package_version;
+use crate::{
+    common::user_agent::{get_package_name, get_platform_info, get_user_agent},
+    error::{ErrorKind, EventHubsError},
+    models::AmqpValue,
+};
+use azure_core::credentials::AccessToken;
+use azure_core::Result;
+use azure_core_amqp::{
+    AmqpClaimsBasedSecurity, AmqpClaimsBasedSecurityApis as _, AmqpConnection,
+    AmqpConnectionApis as _, AmqpConnectionOptions, AmqpSession, AmqpSessionApis as _, AmqpSymbol,
+};
+use std::collections::HashMap;
+use std::sync::{Arc, OnceLock};
+use tokio::sync::Mutex;
+use tracing::{debug, trace};
+use url::Url;
+
+/// The connection manager is responsible for managing the connection to the Event Hubs service.
+/// It also handles authorization and connection recovery.
+///
+/// Currently the connection manager only handles a *single* connection, eventually it will manage
+/// a pool of connections to the service.
+pub(crate) struct ConnectionManager {
+    url: Url,
+    application_id: Option<String>,
+    custom_endpoint: Option<Url>,
+    connections: Mutex<OnceLock<Arc<AmqpConnection>>>,
```

Comment:
More commonly this is a `OnceLock<Mutex<T>>` and you shouldn't need an `Arc` then e.g., <https://github.com/Azure/azure-sdk-for-rust/blob/a8211be103b021d3e0132ca08d1202b170c7f6ce/sdk/core/azure_core_test/src/recording.rs#L57>.

Reply:
The Arc part of this change is forward looking to when this is a connection pool - the idea will be that you can borrow connections from the pool and return them when you're done with the individual API call. That's why the `get()` method returns an `Arc<AmqpConnection>` rather than a `&AmqpConnection`.

And the Mutex is there because `OnceLock::get_or_try_init()` is an experimental feature :(. So I need the mutex to protect the initialization of the OnceLock.

I could just as easily have used `Option<AmqpConnection>` here, to be honest.